File: 04/4b.py
import re
from dataclasses import dataclass


@dataclass
class ScoreRecord:
    game_id: str
    winning_count: int


INPUT_LINES = []
with open("04/input-4a.txt") as infile:
    INPUT_LINES = infile.read().split("\n")

score = 0

all_games = []

for line in INPUT_LINES:
    game_id, all_numbers = line.split(": ")
    winning_str, numbers_str = all_numbers.split(" | ")
    winning = [int(n) for n in re.split("\s+", winning_str.strip())]
    numbers = [int(n) for n in re.split("\s+", numbers_str.strip())]

    common = 0
    for num in numbers:
        if num in winning:
            common += 1

    # print(game_id, common)

    all_games.append(ScoreRecord(game_id=game_id, winning_count=common))

print("====")

all_counts = {}


def incr_count(game_id: str, incr: int):
    if game_id in all_counts:
        all_counts[game_id] += incr
    else:
        all_counts[game_id] = incr


for game_idx, game in enumerate(all_games):
    card_count = 1 + all_counts.get(game.game_id, 0)
    # print(f"Found {card_count} cards of {game.game_id}")

    # print("----", game_idx)
    for idx in range(game_idx + 1, game_idx + 1 + game.winning_count):
        # print(game_idx, idx, all_games[idx].game_id)
        incr_count(all_games[idx].game_id, card_count * 1)

    all_counts[game.game_id] = card_count

# print(all_counts)
print("SCORE", sum(all_counts.values()))

File: 04/typescript/4b.ts
import { readFileSync } from "fs";

const data = readFileSync("04/input-04.txt", { encoding: "utf-8" });

const lines = data.split("\n");

const lineRegex = new RegExp(/:\s+| \| /);
const spaceRegex = new RegExp(/\s+/);

const cardResults: { cardId: string; matches: number }[] = [];
for (const line of lines) {
  const [cardId, winnerStr, numberStr] = line.split(lineRegex);
  const winners = winnerStr.split(spaceRegex).map(Number);
  const numbers = numberStr.split(spaceRegex).map(Number);

  let matches = 0;
  for (const num of numbers) {
    if (winners.includes(num)) {
      matches++;
    }
  }

  cardResults.push({ cardId: cardId, matches: matches });
}

const cardCount = new Map<string, number>();

let resultIdx = 0;
for (const result of cardResults) {
  const numCards = 1 + (cardCount.get(result.cardId) ?? 0);

  for (let idx = resultIdx + 1; idx < resultIdx + 1 + result.matches; idx++) {
    const n = cardCount.get(cardResults[idx].cardId) ?? 0;
    cardCount.set(cardResults[idx].cardId, n + numCards);
  }

  cardCount.set(result.cardId, numCards);

  resultIdx++;
}

let score = 0;

for (const count of cardCount.values()) {
  score += count;
}

console.log("SCORE", score);
